about: Add FAQ page (#57)
Igalia/wpewebkit.org: about/faq.md

---
layout: post
title: "WPE's Frequently Asked Questions"
tags: [about]
data: { dateless: "true" }
permalink: /about/faq.html
---

[[toc]]

## What is the difference between WebPlatformForEmbedded's WPE and WPEWebKit?

Historically the [WebPlatformForEmbedded flavor of
WPE](https://github.com/WebPlatformForEmbedded/WPEWebKit) came first. This

> version includes several adaptations for the RDK platform, as well as different

Reviewer: Would it be fair to say it was developed for RDK? I feel like it would make sense to include something here that said "Reference Design Kit (RDK), a <1 sentence what is that/for">

fixes for its supported devices. The [RDK WPE
page](https://wiki.rdkcentral.com/display/RDK/WPE) provides more information
about WPE in the RDK platform.

> In 2017 the Igalia folks submitted a new flavor of WPE, suitable for upstream

Reviewer: s/the Igalia folks/engineers from Igalia

hosting under the webkit.org umbrella. This version of WPE its released every 6
months from the code hosted on the [upstream repository](https://webkit.org).
This flavor of WPE is maintained upstream and provides regular security updates.

## What Web features does WPE support?

> The WebKit project provides a [features status page](https://webkit.org/status/)

Reviewer: These are things supported by WebKit, but it doesn't necessarily mean they are supported in all ports, right?  Won't some also depend on the backend you choose, at least?

Author: Right, this needs more details as flagged by the TODO.

1. [WPEWebKit upstream has support for WebRTC](https://blogs.gnome.org/tsaunier/2018/07/31/webkitgtk-and-wpe-gains-webrtc-support-back/), by relying on LibWebRTC.
2. LibWebRTC is bundled as third-party library in WebKit's upstream repository
3. The LibWebRTC backend doesn't support hardware-accelerated encoders and decoders
4. LibWebRTC bundles BoringSSL which is a fork of OpenSSL started while OpenSSL
   was still under the dual OpenSSL and SSLeay licences.

Taking in account these facts, the WPEWebKit maintainers have decided to leave
WebRTC support disabled in the default build configuration of the tarballs
because:

- Bundling LibWebRTC in tarballs significantly increases the archive size
- The dependency on BoringSSL prevents LibWebRTC usage in non GPL applications
  and redistribution in binary form
- The lack of hardware-accelerated support in LibWebRTC would incur a bad
  performance impact on the embedded platforms that WPE targets.
  
In order to solve these issues, an alternative WebRTC backend based on
[GstWebRTC](http://blog.nirbheek.in/2018/02/gstreamer-webrtc.html) will be
merged in WPE upstream, most likely in 2021. This new backend will seamlessly

> integrate with hardware-accelerated encoders and decoders on most embedded

Reviewer: ... and address licencing issues, I guess?

## Is Wayland required to run WPE?

As we say in Galicia, "it depends". Read on for more details.

[WPE's architecture](/about/architecture.html) was designed in order to decouple
rendering out of the Web engine and delegate this task to rendering backends and
to the application running the Web engine. So if your application requires a
Wayland compositor, such as Weston, then you need to make sure you use a
Wayland-EGL compatible rendering backend, such as WPEBackend-FDO for instance.
Using this backend's API in your application will allow for smooth rendering in
any Wayland compositor. [Cog](https://github.com/igalia/Cog) is one such
WPE-based application relying on WPEBackend-FDO. If you don't need a Wayland
compositor and just want to render a web-page full-screen, like in most Kiosk
and STB use-cases, then you can render to KMS/DRM (Cog also supports this
use-case).

## Are open dialogs/popups menus supported?

JavaScript popups and dialogs rendering is the duty of the WPE application. The
reference WPE Browser, Cog, has limited support for these features (as of 2020,
cog supports option menus).

## What is this wayland-protocols build dependency about in Cog?

> TODO

Author: OK this is the last one I'd need to address, for the time being.

for instance.

If you are a developer aiming to enable WPE on a certain embedded platform, Cog
combined with WPEBackend-FDO provides the most flexible solution for agile
tinkering and to test WPE's features.

## Is Wayland required to run WPE?

As we say in Galicia, "it depends". Read on for more details.

[WPE's architecture](/about/architecture.html) was designed in order to decouple
rendering out of the Web engine and delegate this task to rendering backends and
to the application running the Web engine. So if your application requires a
Wayland compositor, such as Weston, then you need to make sure you use a
Wayland-EGL compatible rendering backend, such as WPEBackend-FDO for instance.
Using this backend's API in your application will allow for smooth rendering in
any Wayland compositor. [Cog](https://github.com/igalia/Cog) is one such
WPE-based application relying on WPEBackend-FDO. If you don't need a Wayland
compositor and just want to render a web-page full-screen, like in most Kiosk
and STB use-cases, then you can render to KMS/DRM (Cog also supports this
use-case).

## Are open dialogs/popups menus supported?

> JavaScript popups and dialogs rendering is the duty of the WPE application. The
> reference WPE Browser, Cog, has limited support for these features (as of 2020,
> cog supports option menus).

Reviewer 2: I am having some trouble to parse the first phrase here. Also plain HTML like `<input type="file">` can result in a file chooser dialog without involving any JavaScript so I think it's better to remove the name of the scripting language from here :wink:  

For example, how about this:

```suggestion
The application embedding WPE is responsible for rendering popups and dialogs.
The reference WPE Browser, Cog, has limited support for these features (as of 2021,
it supports option menus).
```

## What is the wayland-protocols build dependency about in Cog?

> Depending on which platform rendering plugin is enabled at build-time, the Cog
> browser might depend on the wayland-protocols project in order to generate some
> source files it needs in order to act as a Wayland client to the Compositor
> (server) implementing those protocols.

Reviewer 2: This paragraph has two occurrences of “in order to” quite close to each other, and I think “build-time” does not need the dash. The word “compositor“ does not need to be capitalized.

```suggestion
Depending on which platform rendering plugin is enabled at build time, the Cog
browser might depend on the [wayland-protocols](https://github.com/wayland-project/wayland-protocols) project to generate source files needed in order to act as a Wayland client to the compositor (server) implementing those protocols.
```

So for instance, if you enable the [FDO platform
plugin](https://github.com/Igalia/cog/blob/master/CMakeLists.txt#L57) and want

> to use it at runtime to have Cog running in a Wayland compositor, then the

Reviewer 2: I think's its somewhat clearer to say “Cog running as a Wayland client”, or even “as a Wayland application”:

```suggestion
to use it at runtime to have Cog running as a Wayland application, then the
```

plugin will try to consume some Wayland protocols from the server, such as
`xdg-shell`, `fullscreen-shell-unstable-v1`, `presentation-time` and
`linux-dmabuf-unstable-v1`. Those protocols can't be used without first
generating source files derived from each protocol XML spec definition. This is
all part of the Wayland design.

## Why does the browser/launcher (e.g. Cog) crash at startup?

> If you are building an embedded system image yourself, make sure there is at
> least one font installed that can be used as fallback and that Fontconfig knows
> about it.

Reviewer 2: We can add a tip indicating how to know which fonts are known by Fontconfig:

```suggestion
If you are building an embedded system image yourself, make sure there is at
least one font installed that can be used as fallback by Fontconfig. You can
use the `fc-list` program to print the list of known fonts.
```
